pysindy/feature_library/pde_library.py

```diff
@@ -304,68 +372,42 @@ def transform(self, x):
         """
         check_is_fitted(self)
 
-        x = check_array(x)
-
-        n_samples_full, n_features = x.shape
-        n_samples = n_samples_full // self.num_trajectories
+        xp_full = []
+        for x in x_full:
+            n_features = x.shape[x.ax_coord]
 
-        if float(__version__[:3]) >= 1.0:
-            if n_features != self.n_features_in_:
-                raise ValueError("x shape does not match training shape")
-        else:
-            if n_features != self.n_input_features_:
-                raise ValueError("x shape does not match training shape")
-
-        if np.product(self.grid_dims) > 0:
-            num_time = n_samples // np.product(self.grid_dims)
-        else:
-            num_time = n_samples
-
-        xp_full = np.empty(
-            (self.num_trajectories, n_samples, self.n_output_features_), dtype=x.dtype
-        )
-        if len(self.spatial_grid) > 0:
-            x_full = np.reshape(
-                x,
-                np.concatenate(
-                    [[self.num_trajectories], self.grid_dims, [num_time, n_features]]
-                ),
-            )
-        else:
-            x_full = np.reshape(
-                x,
-                np.concatenate([[self.num_trajectories], [num_time, n_features]]),
-            )
+            if float(__version__[:3]) >= 1.0:
+                if n_features != self.n_features_in_:
+                    raise ValueError("x shape does not match training shape")
+            else:
+                if n_features != self.n_input_features_:
+                    raise ValueError("x shape does not match training shape")
 
-        # Loop over each trajectory
-        for trajectory_ind in range(self.num_trajectories):
-            x = np.reshape(x_full[trajectory_ind], (n_samples, n_features))
-            xp = np.empty((n_samples, self.n_output_features_), dtype=x.dtype)
+            shape = np.array(x.shape)
+            shape[-1] = self.n_output_features_
+            xp = np.empty(shape, dtype=x.dtype)
 
             # derivative terms
-            library_derivatives = np.empty(
-                (n_samples, n_features * self.num_derivatives), dtype=x.dtype
-            )
+            shape[-1] = n_features * self.num_derivatives
+            library_derivatives = np.empty(shape, dtype=x.dtype)
             library_idx = 0
-
             for multiindex in self.multiindices:
-                derivs = np.reshape(
-                    x, np.concatenate([self.grid_dims, [num_time], [n_features]])
-                )
-                for axis in range(self.grid_ndim):
+                derivs = x.copy()
```

Comment:
Can we remove `copy`?

Reply:
@znicolaou  I removed it and all tests passed... If you know a reason for the copy, lmk.

Reply:
I think this change is going to mess up mixed derivatives. I don't think the examples in the notebooks and test contain any fits that require terms like u_{xy}, but I don't think the library terms corresponding to these will be calculated correctly in the current code (I think it will only return the derivative of the last nonzero element of the multiindex). We want to loop over all axes, and accumulate the derivatives for each axis to calculate the full derivative, so the old code first set `derivs=x.copy()`, the calculated `derivs=FiniteDifference(derivs)`, so that each time it loops over an axis, it accumulates a new derivative. My memory is that if you don't include the x.copy(), x itself gets modified during the loop since the assignment is by reference, leading to incorrect results too. I'll explicitly look at the library terms to ensure that they are calculating the right mixed derivatives and push a fix.

Reply:
The only function called on `x` in FiniteDifference._differentiate is `np.full_like` and only attribute access is `x.shape`.  So yes, it's passed by reference, but it shouldn't be mutated.  If it does, that seems like an issue to add in itself.

Reply:
yes, FiniteDifference does not modify x(=derivs), but the assignment `derivs=FiniteDifference(derives)` will then modify x, so that x itself keeps accumulating all the previously calculated derivatives. So we need to make a copy of x each time we want to calculate a mixed derivative

Reply:
Aaah I see why it needs to store the mutation across `for axis in range(self.ind_range)`.  `derivs = x` should work (it does, just verified that it isn't mutated) but now that I understand the copy happens regardless, it's probably fine to make it explicit.

Reply:
Good! I don't remember when I added the .copy() it if I was reaching for straws debugging something, or if there was some case where it was necessary, but I think the way that I got it right was just to plot the derivatives computed by transform for a known input data and verify the mixed derivatives are correct. I'll do a quick check on the code before we finish the review to verify it does what its supposed to. Planning to take a final look at things this weekend, and draft the release notes.

Reply:
The update looks good to me, I ran this quick check:

```
import matplotlib.pyplot as plt
import numpy as np
import pysindy as ps

x=np.arange(128)/128
spatial_grid=np.zeros((128,128,2))
spatial_grid[:,:,0]=x[:,np.newaxis]
spatial_grid[:,:,1]=x[np.newaxis,:]
lib=pde_lib = ps.PDELibrary(derivative_order=2, spatial_grid=spatial_grid)
x=(np.sin(2*np.pi*np.arange(128)[:,np.newaxis]/128)*np.sin(4*np.pi*np.arange(128)[np.newaxis,:]/128))[:,:,np.newaxis]
lib.fit(x)
mat=lib.transform(x)

for i in range(mat.shape[-1]):
    plt.imshow(mat[:,:,i])
    plt.colorbar(label=lib.get_feature_names()[i])
    plt.show()
```
I'll mark this resolved.